pkg/cmd/reverseproxy/cmd.go

```diff
@@ -0,0 +1,227 @@
+	}
+
+	kcmURL, err := url.Parse("https://" + net.JoinHostPort("127.0.0.1", r.kcmPort))
+	if err != nil {
+		return err
+	}
+	cpcURL, err := url.Parse("https://" + net.JoinHostPort("127.0.0.1", r.cpcPort))
+	if err != nil {
+		return err
+	}
+
+	kcmProxy, err := newReverseProxy(kcmURL, r.kcmCABundleFile, kcmPathPrefix)
+	if err != nil {
+		return err
+	}
+	kcmHandler := reverseProxyHandler(kcmProxy)
+
+	cpcProxy, err := newReverseProxy(cpcURL, r.cpcCABundleFile, r.cpcPathPrefix)
+	if err != nil {
+		return err
+	}
+	cpcHandler := reverseProxyHandler(cpcProxy)
+
+	mux := http.NewServeMux()
+
+	mux.HandleFunc(kcmPathPrefix, kcmHandler)
+	mux.HandleFunc(r.cpcPathPrefix, cpcHandler)
+	mux.HandleFunc(r.cpcPathPrefix+"/", cpcHandler)
+
+	srv := &http.Server{Addr: r.bindAddress, Handler: mux}
+	defer srv.Close()
+
+	klog.Infof("Listening on %v", r.bindAddress)
+	return srv.ListenAndServeTLS(r.certFile, r.keyFile)
+}
+
+func (r *reverseProxyOpts) generateSelfSignedCert() error {
+	klog.Warningf("Using insecure, self-signed certificates")
+
+	temporaryCertDir, err := ioutil.TempDir("", "reverse-proxy-serving-cert-")
+	if err != nil {
+		return err
+	}
+	signerName := fmt.Sprintf("reverse-proxy-signer@%d", time.Now().Unix())
+	ca, err := crypto.MakeSelfSignedCA(
+		filepath.Join(temporaryCertDir, "serving-signer.crt"),
+		filepath.Join(temporaryCertDir, "serving-signer.key"),
+		filepath.Join(temporaryCertDir, "serving-signer.serial"),
+		signerName,
+		0,
+	)
+	if err != nil {
+		return err
+	}
+
+	r.certFile = filepath.Join(temporaryCertDir, "tls.crt")
+	r.keyFile = filepath.Join(temporaryCertDir, "tls.key")
+	// nothing can trust this, so we don't really care about hostnames
+	servingCert, err := ca.MakeServerCert(sets.NewString("localhost", "127.0.0.1"), 30)
+	if err != nil {
+		return err
+	}
+	if err := servingCert.WriteCertConfigFile(r.certFile, r.keyFile); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func newReverseProxy(target *url.URL, caBundleFile string, pathPrefix string) (*httputil.ReverseProxy, error) {
+	if target.Path != "" || target.RawPath != "" || target.RawQuery != "" {
+		return nil, errors.New("ReverseProxy Director only supports URL without a path")
+	}
+	proxy := &httputil.ReverseProxy{}
+
+	rootCAs := x509.NewCertPool()
+
+	caCert, err := ioutil.ReadFile(caBundleFile)
+	if err != nil {
+		return nil, err
+	}
+
+	if !rootCAs.AppendCertsFromPEM(caCert) {
+		return nil, errors.New("could not append caCert to root CAs")
+	}
+
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	transport.TLSClientConfig = &tls.Config{
+		RootCAs: rootCAs,
+	}
+	proxy.Transport = transport
+
+	proxy.Director = func(req *http.Request) {
+		req.URL.Scheme = target.Scheme
+		req.URL.Host = target.Host
+		if len(pathPrefix) != 0 && pathPrefix != "/" {
+			req.URL.Path = strings.TrimPrefix(req.URL.Path, pathPrefix)
+			if len(req.URL.RawPath) != 0 {
+				req.URL.RawPath = strings.TrimPrefix(req.URL.EscapedPath(), pathPrefix)
+			}
+		}
```

Comment:
this will need some tests to verify that the pathPrefix and Trim doesn't do something we don't want

I will put it in its own function and unit test it
```go
func trimPathPrefix(pathPrefix string, path string) string {
}
```

Reply:
good idea, I will add this if we decide to proceed with this PR as there are currently concerns a proxy is not a safe approach.

Reply:
that is why I've asked https://github.com/openshift/cluster-kube-controller-manager-operator/pull/615#discussion_r855843178 , I think this will be a common problem that will need a standard solution @deads2k @soltysh